cedar-policy/src/api.rs

```diff
@@ -2247,6 +2303,28 @@ impl Template {
         let est = self.lossless.est()?;
         serde_json::to_value(est).map_err(Into::into)
     }
+
+    /// Get valid [`RequestEnv`]s.
+    /// A [`RequestEnv`] is valid when the template type checks w.r.t requests
+    /// that satisfy it.
+    pub fn get_valid_request_envs(&self, s: &Schema) -> impl Iterator<Item = RequestEnv> {
+        let tc = Typechecker::new(
+            &s.0,
+            cedar_policy_validator::ValidationMode::default(),
+            self.ast.id().clone(),
+        );
+        tc.typecheck_by_request_env(&self.ast)
+            .into_iter()
+            .filter_map(|(env, pc)| {
+                if matches!(pc, PolicyCheck::Success(_)) {
+                    Some(RequestEnv::from(env))
+                } else {
+                    None
+                }
+            })
+            .collect::<BTreeSet<_>>()
+            .into_iter()
```

Comment:
We should be able to avoid collecting into the set here?  Might need to add `+ '_` to the `impl Iterator` return type

Reply:
It seems that I can't avoid collecting because otherwise `tc` will be referenced, which is to a temporary variable.

Reply:
Ok, no problem.